Jinja2Templates assertion error in starlette version >= 0.28

<!-- Please search existing issues to avoid creating duplicates. -->


#### Problem description

- Upgrading to FastAPI 0.108.0, allowed for a starlette upgrade to 0.32.0.post1 (previously I was on 0.27.0).
- Starlette version 0.28 includes [this PR](https://github.com/encode/starlette/pull/2159), which changes the way that `Jinja2Templates()` works.
- The code in `Jinja2Templates.__init__()` now has an `assert directory or env`
- titiler, though, passes in `directory=""` which python evaluates to False, which triggers the assert

```
  ...
  File "./main.py", line 9, in <module>
    from titiler.core.errors import DEFAULT_STATUS_CODES, add_exception_handlers
  File "..../lib/python3.11/site-packages/titiler/core/__init__.py", line 5, in <module>
    from . import dependencies, errors, factory, routing  # noqa
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ..../lib/python3.11/site-packages/titiler/core/factory.py", line 73, in <module>
    DEFAULT_TEMPLATES = Jinja2Templates(
                        ^^^^^^^^^^^^^^^^
  File "..../lib/python3.11/site-packages/starlette/templating.py", line 103, in __init__
    assert directory or env, "either 'directory' or 'env' arguments must be passed"
AssertionError: either 'directory' or 'env' arguments must be passed
```

#### Expected Output

No assertion error.

#### Environment Information

Ubuntu 22.04, python3.11

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -2,9 +2,16 @@
 
 ## Unreleased
 
+### titiler.core
+* fix template loading for starlette >= 0.28 by using `jinja2.Environment` argument
+
+### titiler.extensions
+* fix template loading for starlette >= 0.28 by using `jinja2.Environment` argument
+
 ### titiler.application
 
 * add simple *auth* (optional) based on `global_access_token` string, set with `TITILER_API_GLOBAL_ACCESS_TOKEN` environment variable (author @DeflateAwning, https://github.com/developmentseed/titiler/pull/735)
+* fix template loading for starlette >= 0.28 by using `jinja2.Environment` argument
 
 ## 0.15.6 (2023-11-16)
 

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,21 +1,28 @@
 # Release Notes
 
+## Unreleased
+
+### titiler.core
+
+* fix template loading for starlette >= 0.28 by using `jinja2.Environment` argument
+
+### titiler.extensions
+
+* fix template loading for starlette >= 0.28 by using `jinja2.Environment` argument
+
+### titiler.application
+
+* fix template loading for starlette >= 0.28 by using `jinja2.Environment` argument
+
 ## 0.15.7 (2024-01-08)
 
 ### titiler.core
 
 * update FastAPI version upper limit to `0.107.0` to avoid starlette breaking change (`0.28`)
 
-### titiler.core
-* fix template loading for starlette >= 0.28 by using `jinja2.Environment` argument
-
-### titiler.extensions
-* fix template loading for starlette >= 0.28 by using `jinja2.Environment` argument
-
 ### titiler.application
 
 * add simple *auth* (optional) based on `global_access_token` string, set with `TITILER_API_GLOBAL_ACCESS_TOKEN` environment variable (author @DeflateAwning, https://github.com/developmentseed/titiler/pull/735)
-* fix template loading for starlette >= 0.28 by using `jinja2.Environment` argument
 
 ## 0.15.6 (2023-11-16)
 

--- a/src/titiler/application/titiler/application/main.py
+++ b/src/titiler/application/titiler/application/main.py
@@ -40,10 +40,10 @@ logging.getLogger("botocore.credentials").disabled = True
 logging.getLogger("botocore.utils").disabled = True
 logging.getLogger("rio-tiler").setLevel(logging.ERROR)
 
-templates = Jinja2Templates(
-    directory="",
-    loader=jinja2.ChoiceLoader([jinja2.PackageLoader(__package__, "templates")]),
-)  # type:ignore
+jinja2_env = jinja2.Environment(
+    loader=jinja2.ChoiceLoader([jinja2.PackageLoader(__package__, "templates")])
+)
+templates = Jinja2Templates(env=jinja2_env)
 
 
 api_settings = ApiSettings()
@@ -56,7 +56,8 @@ api_key_query = APIKeyQuery(name="access_token", auto_error=False)
 def validate_access_token(access_token: str = Security(api_key_query)):
     """Validates API key access token, set as the `api_settings.global_access_token` value.
     Returns True if no access token is required, or if the access token is valid.
-    Raises an HTTPException (401) if the access token is required but invalid/missing."""
+    Raises an HTTPException (401) if the access token is required but invalid/missing.
+    """
     if api_settings.global_access_token is None:
         return True
 

--- a/src/titiler/core/pyproject.toml
+++ b/src/titiler/core/pyproject.toml
@@ -29,7 +29,7 @@ classifiers = [
 ]
 dynamic = ["version"]
 dependencies = [
-    "fastapi>=0.100.0,<0.107.0",
+    "fastapi>=0.107.0",
     "geojson-pydantic>=1.0,<2.0",
     "jinja2>=2.11.2,<4.0.0",
     "numpy",

--- a/src/titiler/core/titiler/core/factory.py
+++ b/src/titiler/core/titiler/core/factory.py
@@ -70,11 +70,10 @@ from titiler.core.resources.responses import GeoJSONResponse, JSONResponse, XMLR
 from titiler.core.routing import EndpointScope
 from titiler.core.utils import render_image
 
-DEFAULT_TEMPLATES = Jinja2Templates(
-    directory="",
-    loader=jinja2.ChoiceLoader([jinja2.PackageLoader(__package__, "templates")]),
-)  # type:ignore
-
+jinja2_env = jinja2.Environment(
+    loader=jinja2.ChoiceLoader([jinja2.PackageLoader(__package__, "templates")])
+)
+DEFAULT_TEMPLATES = Jinja2Templates(env=jinja2_env)
 
 img_endpoint_params: Dict[str, Any] = {
     "responses": {

--- a/src/titiler/extensions/titiler/extensions/viewer.py
+++ b/src/titiler/extensions/titiler/extensions/viewer.py
@@ -9,10 +9,10 @@ from starlette.templating import Jinja2Templates
 
 from titiler.core.factory import BaseTilerFactory, FactoryExtension
 
-DEFAULT_TEMPLATES = Jinja2Templates(
-    directory="",
-    loader=jinja2.ChoiceLoader([jinja2.PackageLoader(__package__, "templates")]),
-)  # type:ignore
+jinja2_env = jinja2.Environment(
+    loader=jinja2.ChoiceLoader([jinja2.PackageLoader(__package__, "templates")])
+)
+DEFAULT_TEMPLATES = Jinja2Templates(env=jinja2_env)
 
 
 @dataclass

--- a/src/titiler/extensions/titiler/extensions/wms.py
+++ b/src/titiler/extensions/titiler/extensions/wms.py
@@ -22,10 +22,10 @@ from titiler.core.factory import BaseTilerFactory, FactoryExtension
 from titiler.core.resources.enums import ImageType, MediaType
 from titiler.core.utils import render_image
 
-DEFAULT_TEMPLATES = Jinja2Templates(
-    directory="",
-    loader=jinja2.ChoiceLoader([jinja2.PackageLoader(__package__, "templates")]),
-)  # type:ignore
+jinja2_env = jinja2.Environment(
+    loader=jinja2.ChoiceLoader([jinja2.PackageLoader(__package__, "templates")])
+)
+DEFAULT_TEMPLATES = Jinja2Templates(env=jinja2_env)
 
 
 class WMSMediaType(str, Enum):
@@ -522,7 +522,6 @@ class wmsExtension(FactoryExtension):
                 return image, format, transparent
 
             if request_type.lower() == "getmap":
-
                 # List of required parameters (styles and crs are excluded)
                 req_keys = {
                     "version",

